lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,274 @@
+	Iterator
+	key() []byte
+}
+
+type commonSorted interface {
+	common
+	add(k []byte, v Value)
+	PopFirst(v Value) bool
+	PopLast(v Value) bool
+	newIterator(p iteratorParent, reverse bool) keyIterator
+}
+
+// NewSorted returns an implementaiton of Sorted, spilling to disk at location.
+func NewSorted(location string) Sorted {
+	o := &sorted{base: newBase(location)}
+	o.commonSorted = &memorySorted{}
+	return o
+}
+
+func (o *sorted) Add(k []byte, v Value) {
+	if o.startSpilling(o.Bytes() + v.ProtoSize()) {
+		d, err := v.Marshal()
+		errPanic(err)
+		newSorted := newDiskSorted(o.location)
+		it := o.NewIterator().(keyIterator)
+		for it.Next() {
+			v.Reset()
+			it.Value(v)
+			newSorted.add(it.key(), v)
+		}
+		it.Release()
+		o.commonSorted.Close()
+		o.commonSorted = newSorted
+		o.spilling = true
+		v.Reset()
+		errPanic(v.Unmarshal(d))
+	}
+	o.add(k, v)
+}
+
+func (o *sorted) NewIterator() Iterator {
+	return o.newIterator(false)
+}
+
+func (o *sorted) NewReverseIterator() Iterator {
+	return o.newIterator(true)
+}
+
+func (o *sorted) newIterator(reverse bool) Iterator {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	o.iterating = true
+	return o.commonSorted.newIterator(o, reverse)
+}
+
+// Close is just here to catch deferred calls to Close, such that the correct
+// method is called in case spilling happened.
+func (o *sorted) Close() {
+	o.commonSorted.Close()
+}
+
+func (o *sorted) String() string {
+	return fmt.Sprintf("Sorted@%p", o)
+}
+
+func (o *sorted) released() {
+	o.iterating = false
+}
+
+// memorySorted is basically a slice that keeps track of its size and supports
+// sorted iteration of its element.
+type memorySorted struct {
+	bytes    int
+	outgoing bool
+	values   []Value
+	keys     [][]byte
+}
+
+func (o *memorySorted) add(k []byte, v Value) {
+	if o.outgoing {
+		panic("Add/Append may never be called after PopFirst/PopLast")
+	}
+	o.values = append(o.values, v)
+	o.keys = append(o.keys, k)
+	o.bytes += v.ProtoSize()
+}
+
+func (o *memorySorted) Len() int {
+	return len(o.values)
+}
+func (o *memorySorted) Swap(a, b int) {
+	o.values[a], o.values[b] = o.values[b], o.values[a]
+	o.keys[a], o.keys[b] = o.keys[b], o.keys[a]
+}
+func (o *memorySorted) Less(a, b int) bool {
+	return bytes.Compare(o.keys[a], o.keys[b]) == -1
+}
+
+func (o *memorySorted) Bytes() int {
+	return o.bytes
+}
+
+func (o *memorySorted) Close() {}
+
+func (o *memorySorted) Items() int {
+	return len(o.values)
+}
+
+func (o *memorySorted) PopFirst(v Value) bool {
+	return o.pop(v, true)
+}
+
+func (o *memorySorted) PopLast(v Value) bool {
+	return o.pop(v, false)
+}
+
+func (o *memorySorted) pop(v Value, first bool) bool {
+	if !o.outgoing {
+		sort.Sort(o)
+		o.outgoing = true
+	}
+	if o.Items() == 0 {
+		return false
+	}
+	if first {
+		copyValue(v, o.values[0])
+		o.values = o.values[1:]
+		o.keys = o.keys[1:]
+	} else {
+		i := o.Items() - 1
+		copyValue(v, o.values[i])
+		o.values = o.values[:i]
+		o.keys = o.keys[:i]
+	}
```

Comment:
This could live in it's individual pop function with a preamble maybeSort() called in both. In general, I think this uses the wrong structure.